--- uClibc++-0.2.4/extra/Configs/Config.in.devel~	2012-05-25 23:15:31.000000000 +0200
+++ uClibc++-0.2.4/extra/Configs/Config.in	2012-12-19 02:43:40.713119557 +0100
@@ -260,6 +260,17 @@ config UCLIBCXX_RUNTIME_PREFIX
 	 UCLIBCXX_ RUNTIME_PREFIX is the directory into which the uClibc++
 	  runtime libraries will be installed.
 
+	  For a typical target system this should be set to "/", so
+	  that 'make install' will install libuClibc++.so into
+	  /lib.
+
+config UCLIBCXX_DEVEL_PREFIX
+	string "uClibc++ runtime prefix directory"
+	default "/usr/uClibc++/usr"
+	help
+	 UCLIBCXX_ DEVEL_PREFIX is the directory into which the uClibc++
+	  development environment will be installed.
+
 	  For a typical target system this should be set to "/usr", so
 	  that 'make install' will install libuClibc++.so into
 	  /usr/lib.
--- uClibc++-0.2.4/Rules.mak.devel~	2012-12-19 02:43:40.695119316 +0100
+++ uClibc++-0.2.4/Rules.mak	2012-12-19 02:43:40.714119570 +0100
@@ -90,10 +90,12 @@ BUILD_EXTRA_LIBRARIES:=$(call qstrip,$(B
 
 # Make certain these contain a final "/", but no "//"s.
 UCLIBCXX_RUNTIME_PREFIX:=$(strip $(subst //,/, $(subst ,/, $(call qstrip,$(UCLIBCXX_RUNTIME_PREFIX)))))
+UCLIBCXX_DEVEL_PREFIX:=$(strip $(subst //,/, $(subst ,/, $(call qstrip,$(UCLIBCXX_DEVEL_PREFIX)))))
 UCLIBCXX_RUNTIME_LIBDIR:=$(strip $(subst //,/, $(subst //,/, $(subst ,/, $(call qstrip,$(UCLIBCXX_RUNTIME_PREFIX)$(UCLIBCXX_RUNTIME_LIB_SUBDIR))))))
+UCLIBCXX_DEVEL_LIBDIR:=$(strip $(subst //,/, $(subst //,/, $(subst ,/, $(call qstrip,$(UCLIBCXX_DEVEL_PREFIX)$(UCLIBCXX_RUNTIME_LIB_SUBDIR))))))
 UCLIBCXX_RUNTIME_BINDIR:=$(strip $(subst //,/, $(subst //,/, $(subst ,/, $(call qstrip,$(UCLIBCXX_RUNTIME_PREFIX)$(UCLIBCXX_RUNTIME_BIN_SUBDIR))))))
 UCLIBCXX_RUNTIME_INCLUDEDIR:=$(strip $(subst //,/, $(subst //,/, $(subst ,/, $(call qstrip,$(UCLIBCXX_RUNTIME_PREFIX)$(UCLIBCXX_RUNTIME_INCLUDE_SUBDIR))))))
-export UCLIBCXX_RUNTIME_PREFIX UCLIBCXX_RUNTIME_LIBDIR UCLIBCXX_RUNTIME_BINDIR UCLIBCXX_RUNTIME_INCLUDEDIR
+export UCLIBCXX_RUNTIME_PREFIX UCLIBCXX_DEVEL_PREFIX UCLIBCXX_RUNTIME_LIBDIR UCLIBCXX_DEVEL_LIBDIR UCLIBCXX_RUNTIME_BINDIR UCLIBCXX_RUNTIME_INCLUDEDIR
 
 OPTIMIZATION:=
 PICFLAG:=-fPIC
@@ -139,7 +141,7 @@ GCC_MAJOR_VER?=$(shell $(CC) -dumpversio
 #GCC_MINOR_VER?=$(shell $(CC) -dumpversion | cut -c3)
 
 GEN_LIBS:=
-ifneq ($(LIBGCC_DIR),$(UCLIBCXX_RUNTIME_LIBDIR))
+ifneq ($(LIBGCC_DIR),$(UCLIBCXX_DEVEL_LIBDIR))
 GEN_LIBS += -L$(LIBGCC_DIR)
 endif
 ifneq ($(IMPORT_LIBSUP),y)
--- uClibc++-0.2.4/src/Makefile.devel~	2012-12-19 02:43:40.704119436 +0100
+++ uClibc++-0.2.4/src/Makefile	2012-12-19 04:05:20.918755996 +0100
@@ -25,12 +25,17 @@ all:	libgcc_eh libsupc $(EXOBJS) $(ALLBI
 install:
 	$(INSTALL) -d $(PREFIX)$(UCLIBCXX_RUNTIME_LIBDIR)
 	($(TAR) -cf - \
-	 $(if $(BUILD_STATIC_LIB),$(LIBNAME).a,) \
 	 $(if $(BUILD_ONLY_STATIC_LIB),, \
 		$(SHARED_FULLNAME) \
-		$(SHARED_MAJORNAME) $(LIBNAME).so \
+		$(SHARED_MAJORNAME) \
 	  ) \
 	) | $(TAR) -xf - -C $(PREFIX)$(UCLIBCXX_RUNTIME_LIBDIR)
+ifneq ($(BUILD_ONLY_STATIC_LIB),y)
+	install -m644 $(LIBNAME).so -D $(PREFIX)$(UCLIBCXX_DEVEL_LIBDIR)$(LIBNAME).so
+endif
+ifeq ($(BUILD_STATIC_LIB),y)
+	install -m644 $(LIBNAME).a -D $(PREFIX)$(UCLIBCXX_DEVEL_LIBDIR)$(LIBNAME).a
+endif
 
 libgcc_eh:
 ifeq ($(IMPORT_LIBGCC_EH),y)
@@ -56,10 +61,19 @@ $(LIBNAME).a: libgcc_eh libsupc $(EXOBJS
 	$(AR) $(ARFLAGS) $@ $(EXOBJS) $(GCCOBJS) $(SUPOBJS)
 	$(RANLIB) $@
 
-$(SHARED_FULLNAME): libgcc_eh libsupc $(EXOBJS)
+$(SHARED_FULLNAME): libgcc_eh libsupc $(EXOBJS) $(LIBNAME).so
 	$(CC) $(LDFLAGS) -o $@ $(EXOBJS) $(SUPOBJS) $(GCCOBJS) $(LIBS)
 	$(LN) $@ $(SHARED_MAJORNAME)
-	$(LN) $(SHARED_MAJORNAME) $(LIBNAME).so
+
+OUTPUT_FORMAT = $(CC) $(CFLAGS) -Wl,--verbose 2>&1 | sed -n '/OUTPUT_FORMAT/,/)/p'
+
+$(LIBNAME).so:
+	$(RM) $@
+	$(OUTPUT_FORMAT) >> $@
+	echo "GROUP ( $(PREFIX)$(UCLIBCXX_DEVEL_LIBDIR)libc.so )" >> $@
+	echo "GROUP ( $(PREFIX)$(UCLIBCXX_RUNTIME_LIBDIR)$(SHARED_MAJORNAME) )" >> $@
+	echo "INPUT ( -lgcc )" >> $@
+	echo "INPUT ( AS_NEEDED ( -lgcc_eh ) )" >> $@
 
 clean:
 	$(MAKE) -C abi/libgcc_eh clean
