--- uClibc++-0.2.4/src/Makefile.nostrip~	2012-09-21 02:14:38.717780942 +0200
+++ uClibc++-0.2.4/src/Makefile	2012-09-21 02:14:41.426816142 +0200
@@ -4,9 +4,6 @@ include $(TOPDIR)Rules.mak
 WR_CXX := WRAPPER_INCLUDEDIR=-I$(TOPDIR)include $(TOPDIR)bin/g++-uc
 
 LDFLAGS += -nodefaultlibs -shared -Wl,-soname,$(SHARED_MAJORNAME) $(BUILD_EXTRA_LIBRARIES)
-ifneq ($(DODEBUG),y)
-LDFLAGS += -Wl,-s
-endif
 
 SRCS =	$(wildcard *.cpp)
 SRCS += abi/abi.cpp
