--- uClibc++-0.2.4/src/abi/libsupc/Makefile.atexit_thread~	2014-12-11 21:50:23.619443904 +0100
+++ uClibc++-0.2.4/src/abi/libsupc/Makefile	2014-12-11 21:51:26.679761471 +0100
@@ -15,7 +15,7 @@ libsupc:
 #else
 #	echo Binary
 	$(AR) x $(shell CC="$(CC)" $(TOPDIR)/scripts/find_libsupc.sh)
-	$(RM) -f new_op*.o del_op*.o pure.o new_handler.o eh_alloc.o eh_globals.o
+	$(RM) -f new_op*.o del_op*.o pure.o new_handler.o eh_alloc.o eh_globals.o atexit_thread.o
 #
 #endif
 #endif
